filewatcher/v2/filewatcher.go

```diff
@@ -0,0 +1,567 @@
+		}
+		defer f.Close()
+		reader = f
+	}
+
+	d, err := parser(reader)
+	if err != nil {
+		return zero, time.Time{}, nil, fmt.Errorf("filewatcher: parser error: %w", err)
+	}
+	return d, mtime, files, nil
+}
+
+type opts struct {
+	fsEventsDelay       time.Duration
+	pollingInterval     time.Duration
+	initialReadInterval time.Duration
+
+	fileSizeLimit int64
+}
+
+// Option used in New.
+type Option func(*opts)
+
+// WithFSEventsDelay sets the delay between receiving the fs events and actually
+// reading and parsing the changes.
+//
+// It's used to avoid short bursts of fs events (for example, when watching a
+// directory) causing reading and parsing repetively.
+//
+// Defaut to DefaultFSEventsDelay.
+func WithFSEventsDelay(delay time.Duration) Option {
+	return func(o *opts) {
+		o.fsEventsDelay = delay
+	}
+}
+
+// WithPollingInterval sets the interval to check file changes proactively.
+//
+// Default to DefaultPollingInterval.
+// To disable polling completely, set it to a negative value.
+//
+// Without polling filewatcher relies solely on fs events from the parent
+// directory. This works for most cases but will not work in the cases that
+// the parent directory will be remount upon change
+// (for example, k8s ConfigMap).
+func WithPollingInterval(interval time.Duration) Option {
+	return func(o *opts) {
+		o.pollingInterval = interval
+	}
+}
+
+// WithInitialReadInterval sets the interval to keep retrying to open the file
+// when creating a new FileWatcher, when the file was not initially available.
+//
+// Default to DefaultInitialReadInterval.
+func WithInitialReadInterval(interval time.Duration) Option {
+	return func(o *opts) {
+		o.initialReadInterval = interval
+	}
+}
+
+// WithFileSizeLimit sets the soft file size limit, with the hard limit being
+// 10x (see HardLimitMultiplier) of the set soft limit.
+//
+// This is completely ignored when DirParser is used.
+//
+// If the soft limit is violated,
+// the violation will be reported via slog at error level and prometheus
+// counter of limitopen_softlimit_violation_total,
+// but it does not stop the normal parsing process.
+//
+// If the hard limit is violated,
+// The loading of the file will fail immediately.
+//
+// Default to DefaultMaxFileSize.
+func WithFileSizeLimit(limit int64) Option {
+	return func(o *opts) {
+		o.fileSizeLimit = limit
+	}
+}
+
+// New creates a new FileWatcher.
+//
+// If the path is not available at the time of calling,
+// it blocks until the file becomes available, or context is cancelled,
+// whichever comes first.
+//
+// When logger is non-nil, it will be used to log errors,
+// either returned by parser or by the underlying file system watcher.
+// Please note that this does not include errors returned by the first parser
+// call, which will be returned directly.
+func New[T any](ctx context.Context, path string, parser Parser[T], options ...Option) (*Result[T], error) {
+	opt := opts{
```

Comment:
Could rewrite using the default options and then appending user provided options such as:

```
opt := opts{}
// Default options
optList := []Option{
	WithFSEventsDelay(DefaultFSEventsDelay),
	WithPollingInterval(DefaultPollingInterval),
	WithInitialReadInterval(DefaultInitialReadInterval),
	WithFileSizeLimit(DefaultMaxFileSize),
}
// Default options get overridden by user provided options when processed in order.
optList = append(optList, options...)
```

or using the `WithOptions` note above from @kylelemons 

```
WithOptions(
	// Default options
	WithFSEventsDelay(DefaultFSEventsDelay),
	WithPollingInterval(DefaultPollingInterval),
	WithInitialReadInterval(DefaultInitialReadInterval),
	WithFileSizeLimit(DefaultMaxFileSize),
	// User options
	WithOptions(options...),
)
```